### PR TITLE
fix(graduation-closeout): accept a configurable graduation base

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

`scripts/development-workflow/graduation-closeout.sh` hard-coded `develop` as
the only acceptable graduation PR base, so a nested integration lineage
(wave branch → module branch → `develop`, e.g.
`develop-ventas-e3b` → `develop-sales-module`) failed closeout even though
the module branch is the correct parent for that layer. Reproduced
downstream on `mome-cl/mome-platform` PR #2138.

## Fix

- Add a `--base <branch>` flag to `graduation-closeout.sh` (default:
  `develop`), matching `batch-merge.sh`'s existing `--base` convention
  (per-subcommand/global flag overriding a `TARGET_BASE`-style default,
  same "override for an epic integration branch" framing).
- `validate_graduation_pr` now checks the PR's actual base against the
  configured `$GRADUATION_BASE` instead of a literal `"develop"` string, and
  still fails closed with a clear error when they don't match.
- Added `is_valid_graduation_base`: a real, independent check that `--base`
  itself is `develop` or a `develop-*` integration branch — an arbitrary
  feature/fix branch is rejected up front regardless of what the graduation
  PR happens to target (issue AC-2).
- Audited the rest of the script for other hard-coded `develop` assumptions:
  found and fixed one — the sub-item closing comment
  ("delivered via graduation PR ... to `develop`") now interpolates the
  actual configured base. Also added `GRADUATION_BASE=` to the summary
  output for observability.

## Scope decision: `graduation-closeout-from-merged-pr.sh` (sibling)

Per the task instructions, I read this sibling script closely since it may
share the assumption. It does: `read_merged_graduation_pr` also hard-codes
`base_ref != "develop"`. I deliberately did **not** change it, because:

- It's the merge-time automation fallback, invoked only by
  `.github/workflows/update-tracker-on-merge.yml`, whose trigger is scoped
  to `on.pull_request.branches: [develop]`. A nested graduation PR
  (base = `develop-sales-module`) can **never** cause that workflow to run
  in the first place, regardless of what this script's internal base check
  says — so "fixing" the check alone would be dead code without also
  widening the workflow's trigger scope.
- Protocol 05b Step 5 already documents the manual
  `graduation-closeout.sh` invocation (now with `--base`) as "the primary
  closeout path... Merge-time automation is a fallback only."
- Widening the workflow trigger to catch arbitrary nested integration
  branches raises exactly the kind of "which base does this apply to"
  question that #1329 exists to resolve as documentation of `/run-epic
  --base` semantics. I did not want to silently resolve that scope as a
  side effect of this issue, per the task instructions, so I'm leaving it
  as a documented gap here rather than guessing.

This issue (#1513) is explicitly scoped to "the closeout script's own
validation" per the issue body, distinct from #1329. Both remain open;
#1329 is unaffected.

## Tests

- `scripts/development-workflow/tests/test-graduation-closeout.sh`: 8 new
  tests added, covering:
  - the default `develop` base still works and still rejects a mismatched
    base with a clear error (regression coverage — must not break);
  - `--base develop-sales-module` matching the PR's actual base completes
    closeout normally (nested-graduation happy path);
  - `--base develop-sales-module` with a PR that actually targets `develop`
    is still rejected, with the *configured* base named in the error (not a
    hard-coded `develop`);
  - an invalid `--base` value (e.g. `feature/not-an-integration-branch`) is
    rejected at usage-validation time, before any `gh` call.
- Confirmed all 8 new tests fail against the pre-fix script (stashed the
  script diff, kept the test diff, re-ran — 8 failures, all in the expected
  new tests; original 67 assertions still passed unchanged).
- Full suite after the fix: **75 passed, 0 failed**.
- Sibling suite `test-graduation-closeout-from-merged-pr.sh`: **36 passed, 0
  failed** (unaffected, as expected — that script was not modified).
- `shellcheck` clean on both modified scripts.
- Note: `test-graduation-closeout.sh` is one of the suites CI does not
  execute (#1537) — this local run is the only verification this change
  gets before human review.

## CHANGELOG

Added under `[Unreleased]` → `### Fixed`.

Fixes #1513

## Summary by CodeRabbit

- **New Features**
  - Graduation closeout supports nested integration base branches through `--base`.
  - Defaults to `develop`, with optional environment configuration.
  - Validates supported branch names and confirms the graduation request targets the selected base.
  - Reports the selected base in closeout results and comments.

- **Documentation**
  - Updated development workflow guidance for nested graduations and custom base branches.

- **Tests**
  - Added regression coverage for valid, invalid, default, and mismatched base scenarios.

### CODE REVIEW

**Comment by lhpaul** [comment]
## Code Review (REVIEW.md) — PR #1544

**Scope**: Fast Track fix (issue #1513), reviewed against `REVIEW.md`'s Code Review Checklist (Pass 1 + Pass 2, no spec/plan for Fast Track) and `docs/best-practices/`.

### Verification performed
- `bash scripts/development-workflow/tests/test-graduation-closeout.sh` → **75 passed, 0 failed** (67 pre-existing + 8 new), confirming the PR's test claim.
- `shellcheck` on `graduation-closeout.sh` and `tests/test-graduation-closeout.sh` → clean, no findings.
- `python3 scripts/lint/workflow-shell-guard-lint.py --base-ref origin/develop` → clean.
- Confirmed the `graduation-closeout-from-merged-pr.sh` scope decision independently: `.github/workflows/update-tracker-on-merge.yml` triggers only on `pull_request.branches: [develop]`, so that script's own base check can never see a nested-base graduation PR — the documented rationale for leaving it unchanged is correct.
- Manually traced `is_valid_graduation_base` / `is_valid_slug` edge cases (empty suffix after `develop-`, `development` lookalike, arbitrary feature branch) — all correctly rejected.
- Confirmed all 3 issue #1513 acceptance criteria are covered by the new tests and validation logic.

### Pass 1: Spec Compliance
- Matches the issue's proposed fix and all 3 ACs; no out-of-scope behavior changes to the sibling merge-time script (verified above).
- CHANGELOG entry present under `[Unreleased] → Fixed`, accurate and detailed.
- No stale debug markers, TODOs, or caller inconsistencies found.

### Pass 2: Code Quality
- Option parsing, input validation (`is_valid_graduation_base`, `is_positive_int`, `is_valid_slug`), and error-path messaging are consistent with the rest of the script and the shell-script checklist in `REVIEW.md`.
- Tests are correctly wired to the new mock scenarios (`wrong_base`, `nested_base`) and assert on both the closeout result and the interpolated error/comment text, not just exit codes.

### Finding fixed directly (important — incomplete workflow update)
`docs/workflow/development-workflow/protocols/05b-graduate-development-protocol.md` Step 3 (`gh pr create --base develop`) and Step 5 (`graduation-closeout.sh` invocation) still only showed the `develop`-only examples, with no mention of the new `--base` flag. An operator following the documented protocol for a nested graduation would still open the PR with `--base develop` and run closeout without `--base`, reproducing the exact bug this PR fixes at the script level — even though the CHANGELOG entry already asserted Step 5 was "invoked manually with `--base`" for nested graduations. Pushed a follow-up commit (`docs(graduate-development): document --base for nested graduation`) adding a nested-graduation note to both steps so the documented procedure and the new flag agree.

### Verdict: **APPROVED**

No remaining blocking or important findings. The PR is ready to proceed through the standard external reviewer loop / CI before `ready-for-human-review`.

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), coderabbit (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** coderabbit

**Advisory findings (non-blocking):**
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1544#issuecomment-5365489792))

*Posted automatically by `pr-review-loop.sh`.*

<details>
<summary>Reviewer-loop history (4 iterations)</summary>


```json
{
  "schema": "reviewer_loop_history.v1",
  "pr_number": 1544,
  "updated_at": "2026-08-21T05:29:10Z",
  "history_status": "available",
  "history_unavailable_reason": "",
  "entries": [
    {
      "iteration": 1,
      "recorded_at": "2026-08-21T05:19:15Z",
      "head_sha": "e80bec8d5831b672dde7eb3c1492da33063b2e5e",
      "run_id": "auto-1787289542-33478-3891",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": false,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 2,
      "recorded_at": "2026-08-21T05:21:56Z",
      "head_sha": "e80bec8d5831b672dde7eb3c1492da33063b2e5e",
      "run_id": "auto-1787289582-39930-17375",
      "result": "needs_fixes",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (needs_fixes)"
      ],
      "blocking_count": 1,
      "suggestion_count": 1,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": true,
        "blocking_platform": "coderabbit"
      }
    },
    {
      "iteration": 3,
      "recorded_at": "2026-08-21T05:28:58Z",
      "head_sha": "31f63beaf9945956d92721183691cec5753ecc6a",
      "run_id": "auto-1787289972-19277-5359",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    },
    {
      "iteration": 4,
      "recorded_at": "2026-08-21T05:29:10Z",
      "head_sha": "31f63beaf9945956d92721183691cec5753ecc6a",
      "run_id": "auto-1787290448-82697-13593",
      "result": "clean",
      "reason": "",
      "platforms": [
        "pr-agent (clean)",
        "coderabbit (clean)"
      ],
      "blocking_count": 0,
      "suggestion_count": 0,
      "unresolved_thread_count": 0,
      "late_threads_found": 0,
      "phase_after_clean": {
        "enabled": true,
        "platforms": [
          "coderabbit"
        ],
        "started": true,
        "net_new_blocker": false,
        "blocking_platform": ""
      }
    }
  ]
}
```
</details>

File: docs/workflow/development-workflow/protocols/05b-graduate-development-protocol.md [hunk 1]
```diff
@@ -310,7 +316,11 @@ After the human merges the graduation PR (must use a **merge commit**):
      --epic <epic-issue-number>
    ```
 
-   Add `--exclude-issue <issue-number>` for optional, deferred, cancelled, or explicitly excluded sub-items that must remain open for human disposition. Add `--defer-epic-close` only when the human explicitly requests that the parent epic remain open after the core deliverable graduates.
+   Add `--base <branch>` when this is a nested graduation whose PR base is a
+   parent integration branch rather than `develop` (e.g. `--base
+   develop-sales-module`) — it must match the graduation PR's actual base,
+   and it must be `develop` or a `develop-*` integration branch. Add
+   `--exclude-issue <issue-number>` for optional, deferred, cancelled, or explicitly excluded sub-items that must remain open for human disposition. Add `--defer-epic-close` only when the human explicitly requests that the parent epic remain open after the core deliverable graduates.
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply docs/workflow/development-workflow/protocols/05b-graduate-development-protocol.md:328]: Confirmed — valid finding. Step 5.1 hard-coded `git checkout develop` / `git pull origin develop`, which conflicted with the new `--base` guidance right above it in Step 5.4 for a nested graduation. Fixed in 31f63bea: Step 5.1 now switches to the graduation PR's actual base branch (develop by default, or the parent integration branch for a nested graduation) before deleting the source branch, and the Step 5.4 helper-description bullet no longer claims the PR always targets develop.

File: scripts/development-workflow/tests/test-graduation-closeout.sh [hunk 2]
```diff
@@ -386,6 +391,74 @@ run_contains "validation_reports_unmerged_pr" "has not merged yet" "$validation_
...
+reset_fixture
+export MOCK_GRADUATION_PR_MODE=wrong_base
+export MOCK_NATIVE_MODE=empty
+export MOCK_LABEL_ISSUES="801"
+export MOCK_PR_MODE=none
+write_issue 801 "Default base child" OPEN
+write_issue 900 "Parent epic" OPEN
+set_status 801 "In Development"
+set_status 900 "In Development"
+default_base_reject_result="$(run_closeout)"
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/tests/test-graduation-closeout.sh:409]: Confirmed — valid hygiene improvement. Fixed in 31f63bea: reset_fixture now also unsets GRADUATION_BASE alongside the other MOCK_*/GITHUB_PROJECT_STATUS_* variables, so a caller-set env var can no longer silently change what a default-base test exercises.

**Comment by lhpaul** [comment]
Re: PR-Agent's "Possible Issue" advisory on `is_valid_graduation_base` (auto-acknowledged by pr-review-loop.sh's known false-positive handling for this pattern, same class as issue #511) — declining with evidence:

The finding claims `develop-feature/foo` would pass `is_valid_graduation_base`. It does not. `is_valid_graduation_base` delegates the suffix check to the existing `is_valid_slug`, whose charset allowlist is `[a-zA-Z0-9._-]` — it does **not** include `/`. So `is_valid_slug("feature/foo")` returns 1 (invalid), and `is_valid_graduation_base("develop-feature/foo")` correctly returns 1 (rejected). Verified directly:

```
develop-feature/foo: REJECTED
develop-fix/bar: REJECTED
develop-sales-module: ACCEPTED
develop: ACCEPTED
```

No code change needed.
